diff -rauN a/cJSON/Makefile b/cJSON/Makefile
--- a/cJSON/Makefile	1970-01-01 05:30:00.000000000 +0530
+++ b/cJSON/Makefile	2018-03-06 11:23:02.369929972 +0530
@@ -0,0 +1,20 @@
+#CFLAG  = -Wall -g -O1
+#CFLAG  = -Wall -g -O0
+
+#STRIP  = strip
+
+D_OBJS  = cJSON.o 
+CJSON_LIB  = libcJSON.so 
+LDFLAGS +=  -shared -Wl,-soname,$(CJSON_LIB)
+#-----------------------------------------------------------------------
+all:$(D_OBJS) $(CJSON_LIB)
+
+%.o: %.c
+	$(CC) $(CPPFLAGS) $(CFLAGS) -fPIC -o $@ -c $<
+
+$(CJSON_LIB):$(D_OBJS) 
+	$(CC) -o $(CJSON_LIB)  $^ $(LDFLAGS) -lm 
+
+# Clean up build products.
+clean:
+	rm -f $(D_OBJS) $(CJSON_LIB)
