From 010859b7b0a088cb85d7949167b858c2ae7572e2 Mon Sep 17 00:00:00 2001
From: Ganesh Reddy <ganesh.reddy@nxp.com>
Date: Fri, 1 Jun 2018 18:07:25 +0530
Subject: [PATCH] ENGINE_setup_bsd_cryptodev calling patch

Signed-off-by: Ganesh Reddy <ganesh.reddy@nxp.com>
---
 crypto/evp/c_all.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/crypto/evp/c_all.c b/crypto/evp/c_all.c
index 719e34d..a3ed00d 100644
--- a/crypto/evp/c_all.c
+++ b/crypto/evp/c_all.c
@@ -82,4 +82,9 @@ void OPENSSL_add_all_algorithms_noconf(void)
     OPENSSL_cpuid_setup();
     OpenSSL_add_all_ciphers();
     OpenSSL_add_all_digests();
+#ifndef OPENSSL_NO_ENGINE
+# if defined(__OpenBSD__) || defined(__FreeBSD__) || defined(HAVE_CRYPTODEV)
+    ENGINE_setup_bsd_cryptodev();
+# endif
+#endif
 }
-- 
1.9.1

