diff -rauN a/pppd/plugins/authorize/authorize.c b/pppd/plugins/authorize/authorize.c
--- a/pppd/plugins/authorize/authorize.c	2016-06-16 17:14:31.650936499 +0530
+++ b/pppd/plugins/authorize/authorize.c	2016-06-16 17:14:09.041300485 +0530
@@ -77,6 +77,7 @@
         	while(group[index] != NULL && strlen(group[index])!=0)
                 {
                   free(group[index]);
+                  index++;
                 }
                 return 1;
               }
