--- a/include/linux/if_pppol2tp.h	2018-03-27 16:08:06.455368620 +0530
+++ b/include/linux/if_pppol2tp.h	2018-03-27 15:55:33.527368620 +0530
@@ -74,6 +74,30 @@ enum {
 	PPPOL2TP_MSG_DATA	= (1 << 3),	/* data packets */
 };
 
+/* Structure used to connect() the socket to a particular tunnel UDP 
+    + *  * socket over IPv6. 
+    + *   */ 
+struct pppol2tpin6_addr { 
+    __kernel_pid_t  pid;            /* pid that owns the fd. 
+                                     * 0 => current */ 
+    int     fd;                     /* FD of UDP socket to use */ 
+     
+    __u16 s_tunnel, s_session;      /* For matching incoming packets */ 
+    __u16 d_tunnel, d_session;      /* For sending outgoing packets */ 
+     
+    struct sockaddr_in6 addr;       /* IP address and port to send to */ 
+}; 
+ 
 
-
+struct pppol2tpv3in6_addr { 
+    __kernel_pid_t  pid;            /* pid that owns the fd. 
+                                     * 0 => current */ 
+    int     fd;                     /* FD of UDP or IP socket to use */ 
+     
+    __u32 s_tunnel, s_session;      /* For matching incoming packets */ 
+    __u32 d_tunnel, d_session;      /* For sending outgoing packets */ 
+     
+    struct sockaddr_in6 addr;       /* IP address and port to send to */ 
+}; 
+ 
 #endif
