--- syslog-ng-3.0.8_Orig/src/timeutils.c	2016-01-06 21:04:23.738619411 +0530
+++ syslog-ng-3.0.8/src/timeutils.c	2016-01-06 21:04:46.326619779 +0530
@@ -229,7 +229,7 @@ format_zone_info(gchar *buf, size_t bufl
   return g_snprintf(buf, buflen, "%c%02ld:%02ld",
                           gmtoff < 0 ? '-' : '+',
                           (gmtoff < 0 ? -gmtoff : gmtoff) / 3600,
-                          (gmtoff % 3600) / 60);
+                          (abs(gmtoff) % 3600) / 60);
 }
 
 /**
