From a6c4f52a54df40a17437cc31c5ae0f6d2b6238c5 Mon Sep 17 00:00:00 2001
From: Rajeshkumar k <rajeshkumar.k@nxp.com>
Date: Mon, 23 Apr 2018 20:50:09 +0530
Subject: [PATCH] Console disable patch for baud rate 9600.

---
 drivers/tty/serial/8250/8250_core.c | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/drivers/tty/serial/8250/8250_core.c b/drivers/tty/serial/8250/8250_core.c
index 4506e40..b0b90ab 100644
--- a/drivers/tty/serial/8250/8250_core.c
+++ b/drivers/tty/serial/8250/8250_core.c
@@ -1531,7 +1531,10 @@ ignore_char:
 		lsr = serial_in(up, UART_LSR);
 	} while ((lsr & (UART_LSR_DR | UART_LSR_BI)) && (--max_count > 0));
 	spin_unlock(&port->lock);
-	tty_flip_buffer_push(&port->state->port);
+
+        if(console_set_on_cmdline) {
+		tty_flip_buffer_push(&port->state->port);
+        }
 	spin_lock(&port->lock);
 	return lsr;
 }
@@ -1544,7 +1547,13 @@ void serial8250_tx_chars(struct uart_8250_port *up)
 	int count;
 
 	if (port->x_char) {
+
+          if(console_set_on_cmdline){
 		serial_out(up, UART_TX, port->x_char);
+          }
+          else{
+		serial_out(up, UART_TX, 0);
+          }
 		port->icount.tx++;
 		port->x_char = 0;
 		return;
@@ -1560,7 +1569,13 @@ void serial8250_tx_chars(struct uart_8250_port *up)
 
 	count = up->tx_loadsz;
 	do {
+          if(console_set_on_cmdline){
 		serial_out(up, UART_TX, xmit->buf[xmit->tail]);
+          }
+          else{
+		serial_out(up, UART_TX, 0);
+          }
+
 		xmit->tail = (xmit->tail + 1) & (UART_XMIT_SIZE - 1);
 		port->icount.tx++;
 		if (uart_circ_empty(xmit))
@@ -3358,6 +3373,10 @@ static void serial8250_console_write(struct uart_8250_port *up, const char *s,
 	unsigned int ier;
 	int locked = 1;
 
+        if(!console_set_on_cmdline){
+          return;
+        }
+
 	touch_nmi_watchdog();
 
 	serial8250_rpm_get(up);
-- 
2.7.4

