--- a/linux/include/linux/if_pppol2tp.h	2018-03-27 16:01:47.287368619 +0530
+++ b/linux/include/linux/if_pppol2tp.h	2018-03-27 16:04:22.527368622 +0530
@@ -36,6 +36,21 @@ struct pppol2tp_addr
 	__u16 d_tunnel, d_session;	/* For sending outgoing packets */
 };
 
+/* Structure used to connect() the socket to a particular tunnel UDP
+    + *  * socket over IPv6.
+    + *   */
+struct pppol2tpin6_addr {
+    __kernel_pid_t  pid;            /* pid that owns the fd.
+                                     * 0 => current */
+    int     fd;                     /* FD of UDP socket to use */
+    
+    __u16 s_tunnel, s_session;      /* For matching incoming packets */
+    __u16 d_tunnel, d_session;      /* For sending outgoing packets */
+    
+    struct sockaddr_in6 addr;       /* IP address and port to send to */
+    };
+
+
 /* The L2TPv3 protocol changes tunnel and session ids from 16 to 32
  * bits. So we need a different sockaddr structure.
  */
@@ -50,6 +65,17 @@ struct pppol2tpv3_addr {
 	__u32 d_tunnel, d_session;	/* For sending outgoing packets */
 };
 
+struct pppol2tpv3in6_addr {
+    __kernel_pid_t  pid;            /* pid that owns the fd.
+                                     * 0 => current */
+    int     fd;                     /* FD of UDP or IP socket to use */
+    
+    __u32 s_tunnel, s_session;      /* For matching incoming packets */
+    __u32 d_tunnel, d_session;      /* For sending outgoing packets */
+    
+    struct sockaddr_in6 addr;       /* IP address and port to send to */
+    };
+
 /* Socket options:
  * DEBUG	- bitmask of debug message categories
  * SENDSEQ	- 0 => don't send packets with sequence numbers
