--- a/netfilter/iptcrdr.c	2017-01-25 12:26:17.955326197 +0530
+++ b/netfilter/iptcrdr.c	2017-01-25 12:23:06.143466549 +0530
@@ -1089,6 +1089,11 @@ addnatrule(int proto, unsigned short epo
         unsigned long current_rule_count;
 
 
+        if(eport == 4500 || eport == 500)
+        {
+            syslog(LOG_ERR,"This port %d is not allowed to use\n",eport);
+            return -1;
+        }
        /* check for the max rule,and return */
         current_rule_count = get_current_chain_rule_count(miniupnpd_nat_chain);
         if(max_allowed_rules == current_rule_count )
@@ -1326,6 +1331,7 @@ add_filter_rule(int proto, const char *
 	struct ipt_entry_match *match = NULL;
 	struct ipt_entry_target *target = NULL;
 
+
 	e = calloc(1, sizeof(struct ipt_entry));
 	if(!e) {
 		syslog(LOG_ERR, "%s: calloc(%d) error", "add_filter_rule",
