diff -Naur barebox-2011.06.0-old/arch/arm/boards/comcerto-rv16x-26x/env/config barebox-2011.06.0/arch/arm/boards/comcerto-rv16x-26x/env/config
--- barebox-2011.06.0-old/arch/arm/boards/comcerto-rv16x-26x/env/config	2017-08-09 13:32:37.291363445 +0530
+++ barebox-2011.06.0/arch/arm/boards/comcerto-rv16x-26x/env/config	2017-08-09 13:47:34.267363338 +0530
@@ -57,7 +57,7 @@
 fast_spi_parts="128k(uloader)ro,512k(barebox)ro,128k(env),128k(boardinfo)ro,64k(md5sum1),64k(md5sum2),64k(nvram)"
 
 nand_device="comcertonand"
-nand_parts="1024k(barebox),1024k(bareboxfact)ro,1024k(env),8M(kernel1),92M(rootfs1),1024k(reserved_dtb1),8M(kernel2),92M(rootfs2),1024k(reserved_dtb2),2M(configcert),12M(reserved),32M(webrootdb),2M(license)"
+nand_parts="1M(barebox),1M(bareboxfact)ro,1M(env),8M(kernel1),140M(rootfs1),1M(reserved_1),8M(kernel2),140M(rootfs2),1M(reserved_2),20M(configcert),12M(config_bkup),48M(webrootdb),8M(license),50M(packages),48M(findIT),24M(reserved)"
 
 autoboot_timeout=3
 
