--- a/arch/arm/mach-comcerto/board-c2krv16x-26x.c	2017-0720 02:30:21.136858871 -0500
+++ b/arch/arm/mach-comcerto/board-c2krv16x-26x.c	2017-07-20 02:46:36.236903977 -0500
@@ -1006,7 +1006,7 @@
 	.init_machine	= platform_init,
 	.init_time	= comcerto_timer_init,
 #ifdef CONFIG_ZONE_DMA
-	.dma_zone_size	= SZ_32M + 3*SZ_4M,
+	.dma_zone_size	= SZ_32M + 11*SZ_4M,
 #endif
 	.restart	= ls1024_restart,
 	.reboot_mode	= REBOOT_SOFT,
--- a/arch/arm/Makefile	2017-07-20 01:40:51.932721521 -0500
+++ b/arch/arm/Makefile	2017-07-20 01:40:59.272721860 -0500
@@ -151,7 +151,7 @@
 machine-$(CONFIG_ARCH_CNS3XXX)		+= cns3xxx
 machine-$(CONFIG_ARCH_COMCERTO)         := comcerto
 ifeq ($(CONFIG_ARCH_COMCERTO),y)
-textofs-$(CONFIG_ZONE_DMA) := 0x04008000
+textofs-$(CONFIG_ZONE_DMA) := 0x06008000
 endif
 machine-$(CONFIG_ARCH_DAVINCI)		+= davinci
 machine-$(CONFIG_ARCH_DIGICOLOR)	+= digicolor
--- a/arch/arm/mach-comcerto/Makefile.boot	2017-07-20 01:35:05.972705517 -0500
+++ b/arch/arm/mach-comcerto/Makefile.boot	2017-07-20 01:35:09.924705700 -0500
@@ -1,6 +1,6 @@
 ifeq ($(CONFIG_ARCH_M86XXX),y)
 ifeq ($(CONFIG_COMCERTO_ZONE_DMA_NCNB),y)
-   zreladdr-y     := 0x04008000
+   zreladdr-y     := 0x06008000
 else
    zreladdr-y     := 0x00008000
 endif
--- a/arch/arm/mach-comcerto/include/mach/comcerto-2000.h	2017-07-20 01:41:42.064723840 -0500
+++ b/arch/arm/mach-comcerto/include/mach/comcerto-2000.h	2017-07-20 01:41:48.612724143 -0500
@@ -119,7 +119,7 @@
 
 #define COMCERTO_AXI_ACP_SIZE			(1 << 24)
 
-#define COMCERTO_DDR_SHARED_BASE		(COMCERTO_AXI_DDR_BASE + 0x2C00000)
+#define COMCERTO_DDR_SHARED_BASE		(COMCERTO_AXI_DDR_BASE + 0x4C00000)
 #define COMCERTO_DDR_SHARED_SIZE		(SZ_16M + SZ_4M)
 #define COMCERTO_FASTSPI_DDR_BASE		(COMCERTO_AXI_DDR_BASE + 0x2B00000)
 #define COMCERTO_FASTSPI_DDR_SIZE		SZ_64K
